### PR TITLE
Register agent API support for MLClient

### PR DESCRIPTION
**Pull request by arjunkumargiri**
### Description
Register agent API support for MLClient. This will allow flow framework to integrate with register agent API.
 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java [hunk 1]
```diff
@@ -253,6 +256,16 @@ public void getTool(String toolName, ActionListener<ToolMetadata> listener) {
         client.execute(MLGetToolAction.INSTANCE, mlToolGetRequest, getMlGetToolResponseActionListener(listener));
     }
 
+    @Override
+    public void registerAgent(MLRegisterAgentRequest mlRegisterAgentRequest, ActionListener<MLRegisterAgentResponse> listener) {
+        client
+            .execute(
+                MLRegisterAgentAction.INSTANCE,
+                mlRegisterAgentRequest,
+                ActionListener.wrap(listener::onResponse, listener::onFailure)
```
Comment (ylwu-amzn) [inline client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: I think this way will fail if calling from AI flow side. Can you verify?

Comment (arjunkumargiri) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: Can you please provide more details on why this would fail?

@dbwiddis can you please check if this mechanism would fail?

Comment (dbwiddis) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: @arjunkumargiri haven't checked yet, but "probably".

See https://github.com/opensearch-project/ml-commons/pull/1616/files for the pattern to use to avoid using different class loaders for the Action listener.

CC: @owaiskazi19 @joshpalis

Comment (joshpalis) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: I would suggest following this [commit](https://github.com/opensearch-project/ml-commons/commit/882246c04daacc98eb9c72fbdcabcce497856423) instead, since the `GetMLTask` API responds with an `MLTask` rather than a response object

Comment (dbwiddis) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: > Can you please provide more details on why this would fail?

See https://github.com/opensearch-project/ml-commons/issues/1553 for reference and other examples.

I'm not sure of the exact inner workings of the JVM causing this, and it may have to do with which order the plugins are loaded on bootstrap, but the pattern you use can cause the Response class to be loaded from a different class loader than the one declared in flow framework which is calling this method, and somehow creating a separate method manages to fix it.

TLDR: in `client.execute()` use a getter method to get the action listener instead of declaring it inline.

Comment (arjunkumargiri) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: I would prefer to return the response object as it is possible that other attributes are added to response object in future. Do we know what flow framework is loading duplicate jars which is leading to class cast exception?

Comment (owaiskazi19) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:265]: @arjunkumargiri replied to the issue. Also, yes this would fail on flow framework side and to avoid we should follow the similar pattern as this PR: https://github.com/opensearch-project/ml-commons/pull/1560/files

File: client/src/main/java/org/opensearch/ml/client/MachineLearningClient.java [hunk 2]
```diff
@@ -337,4 +339,22 @@ default ActionFuture<ToolMetadata> getTool(String toolName) {
      * @param listener action listener
      */
     void getTool(String toolName, ActionListener<ToolMetadata> listener);
+
+    /**
+     * Registers new agent and returns ActionFuture.
+     * @param mlRegisterAgentRequest Register agent input, refer: https://opensearch.org/docs/latest/ml-commons-plugin/api/#register-agent
+     * @return the result future
+     */
+    default ActionFuture<MLRegisterAgentResponse> registerAgent(MLRegisterAgentRequest mlRegisterAgentRequest) {
```
Comment (owaiskazi19) [inline client/src/main/java/org/opensearch/ml/client/MachineLearningClient.java:348]: Based on the previous patterns of the APIs, we generally create the request in `MachineLearningNodeClient`
```suggestion
    default ActionFuture<MLRegisterAgentResponse> registerAgent(MLAgent mlAgent) {
```

File: client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java [hunk 3]
```diff
@@ -253,6 +256,16 @@ public void getTool(String toolName, ActionListener<ToolMetadata> listener) {
         client.execute(MLGetToolAction.INSTANCE, mlToolGetRequest, getMlGetToolResponseActionListener(listener));
     }
 
+    @Override
+    public void registerAgent(MLRegisterAgentRequest mlRegisterAgentRequest, ActionListener<MLRegisterAgentResponse> listener) {
+        client
```
Comment (owaiskazi19) [inline client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:261]: We can create a `MLRegisterAgentRequest` request here by using the `MLAgent` passed from the `MachineLearningClient`

File: client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java (hunk 1 above)
Comment (arjunkumargiri) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:267]: Updated PR to unblock flow integration. However, it will be good to identify why duplicate jars are pulled into flow framework?

**Review by owaiskazi19 — APPROVED** [review]
LGTM

**Review by ylwu-amzn — APPROVED** [review]
LGTM. Thanks a lot for the change

**Comment by ylwu-amzn** [comment]
BTW, the Github CI failed, maybe need to rebase ?

File: client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java (hunk 1 above)
Comment (dbwiddis) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:267]: > Updated PR to unblock flow integration. However, it will be good to identify why duplicate jars are pulled into flow framework?

I think I answered this on another issue but I will repeat it here.
- there is no duplicate JAR, there are two different class loaders involved
- OpenSearch bootstraps plugins with their own class loader which is a child of the OpenSearch parent class loader.  This keeps plugin classloaders isolated from each other.
- so two different Java `.class` objects are created in memory representing the identical compiled Java class from the JAR.  When they are passed between plugins the JVM thinks they are being cast to each other.

File: client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java (hunk 3 above)
Comment (arjunkumargiri) [reply client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java:263]: Done.
